geode-core/src/test/java/org/apache/geode/internal/cache/PartitionedRegionTest.java

```diff
@@ -573,6 +578,50 @@ public void transactionThrowsTransactionDataRebalancedExceptionIfIsAForceReattem
         .hasMessage(PartitionedRegion.DATA_MOVED_BY_REBALANCE).hasCause(exception);
   }
 
+  @Test
+  public void failuresSavedIfFetchKeysThrows() throws Exception {
+    PartitionedRegion spyPartitionedRegion = spy(partitionedRegion);
+
+    VersionedObjectList values = mock(VersionedObjectList.class);
+    ServerConnection serverConnection = mock(ServerConnection.class);
+    Set<Integer> failures = mock(Set.class);
+    InternalDistributedMember member = mock(InternalDistributedMember.class);
+    Set<Integer> buckets = new HashSet<>();
+    buckets.add(1);
+    doThrow(new ForceReattemptException("")).when(spyPartitionedRegion).getFetchKeysResponse(member,
+        1);
+
+    spyPartitionedRegion.fetchKeysAndValues(values, serverConnection, failures, member, null,
+        buckets);
+
+    verify(failures).add(1);
+    verify(spyPartitionedRegion, never()).getValuesForKeys(values, serverConnection, null);
+  }
+
+  @Test
+  public void fetchKeysAndValuesInvokesGetValuesForKeys() throws Exception {
+    PartitionedRegion spyPartitionedRegion = spy(partitionedRegion);
+
+    VersionedObjectList values = mock(VersionedObjectList.class);
+    ServerConnection serverConnection = mock(ServerConnection.class);
+    Set<Integer> failures = mock(Set.class);
+    InternalDistributedMember member = mock(InternalDistributedMember.class);
+    Set<Integer> buckets = new HashSet<>();
+    buckets.add(1);
+    FetchKeysMessage.FetchKeysResponse fetchKeysResponse =
+        mock(FetchKeysMessage.FetchKeysResponse.class);
+    doReturn(fetchKeysResponse).when(spyPartitionedRegion).getFetchKeysResponse(member, 1);
+    Set keys = mock(Set.class);
```

Comment:
Here you should use a real Set instead of a mock.

Reply:
I'd like to know the reason why this has to be a real Set.